### ISSUE TITLE
Sage Maker

### ISSUE THREAD

**Issue by yoonghan** [yoonghan/yoonghan.github.io#519]
## Sage Maker Clarify

** Involve Human **
Use to evaluate Foundation Model based on own/built-in dataset. Let you choose between Model A/Model B on relevance and how it should voice output.
You can have AWS staff/Turk or Own people to evaluate the result.
Part of Sage Maker Studio.

** Model Explanation **
Contains a set of tools for Sage Maker Clarify to **explain** the result. You can debug the result.
Example asking "Why did the model predict negative outcome", it will show a graph of all the fields involved that make the result negative.

** Detect Bias **
To detect and explain Biases in model and dataset.
Specify input features and biases will be detected.

<img width="404" alt="Image" src="https://github.com/user-attachments/assets/efa94e47-a190-4402-9cc4-d94132ba4017" />

**Comment by yoonghan** [comment]
Sage Maker Ground Truth

For reinforcement learning (RLHF) where human feedback is the "reward".

Sage Maker Ground Truth +(PLUS) = For data labelling.

**Comment by yoonghan** [comment]
## Governance

SageMaker Model Card = model information - intended use, risk ratings and training details
SageMaker Model Dashboard = centralised repository - information and insights of all models, which is deployed, which break threshold
SageMaker Model Dashboard  -> Model Monitor = show model Bias/Quality, can setup a monitor/alert
SageMaker Role Manager = define roles and permissions for personas - data scientist, MLOps engineer

Model Registry = Like our artifact but store model
SageMaker Pipeline = It's a CI/CD for deploying a model to production. Supported step type.
   - Processing
   - Training
   - Tuning
   - AutoML
   - Model
   - ClarifyCheck
   - QualityCheck

**Comment by yoonghan** [comment]
## SageMaker Jumpstart

**Contains Open Source FM**

Like Amazon Bedrock FM selection, it's a marketplace. Split to
  - ML Hub = Browse -> Experiment -> Customize -> Deploy
  - ML Solutions = a Saas, Browse -> Select & Customize -> Deploy


Your inference and training data will be used to train the base model - Your inference and training data will not be used nor shared to update or train the base model that SageMaker JumpStart surfaces to customers.

**Comment by yoonghan** [comment]
## SageMaker Canvas

No code interface 
Ready to use/integrate models in Jumpstart or Bedrock
Model is code with *AutoML* powered by *SageMaker Autopilot*
Leverage *Data Wrangler* for data preparation

**Comment by yoonghan** [comment]
## MLFlow
An opensource to manage lifecycle. Used to track, organize, view, analyze, and compare iterative ML experimentation to gain comparative insights and register and deploy your best-performing models. Can integrate with SageMaker.

MLFlow Tracking Servers = track run & experiments

**Comment by yoonghan** [comment]
## Extra Feature

Network Isolation Mode - Really no external network including Amazon S3
Sage Maker DeepAR = forecasting time series data

**Comment by yoonghan** [comment]
Summary

| Tool | Description |
| --- | --- |
| SageMaker Automatic Model Tuning | tune hyperparameters. If Q ask anything to tune = NONE. |
| SageMaker Deployment & Interference | real-time, serverless, batch / async |
| SageMaker Studio | UI for Sage Maker |
| SageMaker Data Wrangler | explore/prepare datasets, create features. *Can integrate with SQL* |
| SageMaker Feature Store | store features metadata in central place. *Data Wrangler* can publish data to Feature Store. Variables can also be stored here. |
| SageMaker Clarify | compare models, explain model, detect bias |
| SageMaker Ground Truth | RLHF for model grading. Has built-in workflows for labeling, supports private/workforce teams for sensitive data (e.g., medical images), and integrates with S3. |
| SageMaker Model Cards | ML model documentation. The intended uses of a model go beyond technical details and describe how a model should be used in production. |
| SageMaker Model Dashboard | view all model in one dashboard |
| SageMaker Model Monitor | Monitor and alert model, this is not Dashboard |
| SageMaker Model Registry | Model registry |
| SageMaker Pipeline | Ci/Cd for Machine Language |
| SageMaker Role Manager | Access and permission control. A baseline set of permissions for ML activities and personas through a catalog of prebuilt AWS Identity and Access Management (IAM) policies. |
| SageMaker JumpStart | ML model hub & prebuilt solution |
| SageMaker Canvas | No-code interface |
| MLFlow for SageMaker | MLFlow tracking servers |

**Comment by yoonghan** [comment]
## SageMaker - Model Deployment & Inference

https://docs.aws.amazon.com/sagemaker/latest/dg/how-it-works-deployment.html

Deploy serverless with autoscale, but have 4 options:
* Real-time - one prediction at a time with setup ram/gpu
* Serverless - idle period between traffic-spikes with ram, can tolerate more latency (cold starts). Ideal if real-time is not predictable.
* Asynchronous
  * Up-to 1GB
  * For long processing 
* Batch
   * Best if above 1GB (see async)
   * Multiple

For async/batch - all these are put into staging bucket/S3 to process and result in S3 in job queue

A SageMaker Inference endpoint allows clients to invoke deployed models and receive predictions

<img width="731" alt="Image" src="https://github.com/user-attachments/assets/1b363bc6-9775-4dac-9978-b63f65e9b54d" />

**Comment by yoonghan** [comment]
A VPC endpoint for Amazon S3 is the most appropriate choice because it creates a private connection between the VPC and Amazon S3 over the AWS network, without requiring internet access. This VPC endpoint allows the SageMaker model deployed within the VPC to securely access data from Amazon S3 directly, using the internal AWS network paths. It provides enhanced security by keeping data traffic within the AWS infrastructure and not exposing it to the public internet.
You can use two types of VPC endpoints to access Amazon S3: gateway endpoints and interface endpoints (by using AWS PrivateLink). A gateway endpoint is a gateway that you specify in your route table to access Amazon S3 from your VPC over the AWS network. Interface endpoints extend the functionality of gateway endpoints by using private IP addresses to route requests to Amazon S3 from within your VPC, on premises, or from a VPC in another AWS Region by using VPC peering or AWS Transit Gateway.

**Comment by yoonghan** [comment]
Amazon SageMaker Studio offers a broad set of fully managed integrated development environments (IDEs) for ML development
- JupyterLab
- Code Editor based on Code-OSS (Visual Studio Code – Open Source)
- RStudio.

**Comment by yoonghan** [comment]
Amazon SageMaker Model Cards have a defined structure that cannot be modified. This structure gives guidance on what information should be captured in a model card.
